File: 0012/P0012.py
import math

from typing import List, Tuple

def calculate_primes(n: int) -> List[int]:
    is_prime = [True for i in range(n+1)]
    is_prime[0] = False
    is_prime[1] = False
    for i in range(2, n+1):
        if is_prime[i]:
            for j in range(2 * i, n+1, i):
                is_prime[j] = False
    return [i for i in range(n+1) if is_prime[i]]

def prime_factors(n: int, primes: List[int]) -> List[Tuple[int, int]]:
    factors = []
    for prime in primes:
        power = 0
        while n % prime == 0:
            n /= prime
            power += 1
        if power > 0:
            factors.append((prime, power))
        if n == 1:
            break
    return factors

def fast_count_factors(n: int, primes: List[int]) -> int:
    factors = prime_factors(n, primes)
    count = 1
    for (_factor, power) in factors:
        count *= power + 1
    return count

def calculate_triangles():
    n = 1
    while True:
        yield n * (n + 1) / 2
        n += 1

if __name__ == '__main__':
    primes = calculate_primes(100)
    cache = [None for _ in range(1001)]
    triangles = calculate_triangles()
    triangle = 0
    factors = 0
    for i in range(1, 1001):
        while True:
            if factors > i:
                cache[i] = triangle
                break
            else:
                triangle = next(triangles)
                factors = fast_count_factors(triangle, primes)

    t = int(input())
    for _ in range(t):
        n = int(input())
        print(int(cache[n]))

File: 0012/P0012.cpp
#include <cmath>
#include <iostream>
#include <vector>
#include <utility>

std::vector<int> calculate_primes(int n) {
  std::vector<int> primes;
  std::vector<bool> is_prime(n+1);

  for (int i = 0; i < is_prime.size(); ++i) {
    is_prime[i] = true;
  }

  is_prime[0] = false;
  is_prime[1] = false;

  for (int i = 2; i < is_prime.size(); ++i) {
    if (is_prime[i]) {
      primes.push_back(i);
      for (int j = i + i; j < is_prime.size(); j += i) {
        is_prime[j] = false;
      }
    }
  }

  return primes;
}

std::vector<std::pair<int, int>> prime_factors(int n, const std::vector<int>& primes) {
  std::vector<std::pair<int, int>> factors;

  for (int prime : primes) {
    int power = 0;
    while (n % prime == 0) {
      n /= prime;
      ++power;
    }

    if (power > 0) {
      std::pair<int, int> factor(prime, power);
      factors.push_back(factor);
    }
    
    if (n == 1) {
      break;
    }
  }

  return factors;
}

int fast_count_factors(int n, const std::vector<int>& primes) {
  const std::vector<std::pair<int, int>> factors = prime_factors(n, primes);

  int count = 1;
  for (const std::pair<int , int>& factor_and_power : factors) {
    count *= factor_and_power.second + 1;
  }

  return count;
}

int calculate_triangle(int n) {
  return n * (n + 1) / 2;
}

int main() {
  const std::vector<int> primes = calculate_primes(100);
  std::vector<int> cache(1001);
  int n = 0;
  int triangle = 0;
  int factors = 0;
  for (int i = 1; i < 1001; ++i) {
    while (true) {
      if (factors > i) {
        cache[i] = triangle;
	break;
      } else {
	++n;
        triangle = calculate_triangle(n);
	factors = fast_count_factors(triangle, primes);
      }
    }
  }

  int t = 0;
  std::cin >> t;
  for (int i=0; i < t; ++i) {
    int n = 0;
    std::cin >> n;
    std::cout << cache[n] << std::endl;
  }
}
